index.js

```diff
@@ -146,27 +133,57 @@ module.exports.bootstrap = async ({
 };
 
 module.exports.transform = ({ data, getPluginContext, options }) => {
+  // If preview is not explicitly set, we default to `true` when in watch mode.
+  const isPreview =
+    options.preview !== undefined ? options.preview : options.watch;
   const { entries = {} } = getPluginContext();
-  const normalizedEntries = normalizeEntries({ entries, options });
-  const models = {};
 
-  normalizedEntries.forEach(entry => {
-    const { __metadata: meta, ...fields } = entry;
+  // Merging published entries with drafts.
+  const mergedEntries = Object.keys(entries).reduce((result, entryId) => {
+    const { canonicalId, isDraft } = parseEntryId(entryId);
 
-    if (meta.modelName.indexOf("system.") === 0) return;
+    // We discard this entry if it's a draft and preview mode is disabled, or
+    // if there's already an entry for this canonical ID and the entry we are
+    // processing now is not a draft (if it is, it takes precedence).
+    if ((isDraft && !isPreview) || (result[canonicalId] && !isDraft)) {
+      return result;
+    }
 
-    models[meta.modelName] = models[meta.modelName] || {
-      fieldNames: Object.keys(fields),
-      modelName: meta.modelName,
-      projectEnvironment: meta.projectEnvironment,
-      projectId: meta.projectId,
-      source: meta.source
+    return {
+      ...result,
+      [canonicalId]: entries[entryId]
     };
+  }, {});
+
+  // Normalizing entries and resolving links. They're also converted to an
+  // array.
+  const normalizedEntries = normalizeEntries({
+    entries: mergedEntries,
+    options
   });
 
+  // Iterating through the entries to compile a hash of models and normalize
+  // it.
+  const models = normalizedEntries.reduce((result, entry) => {
```

Comment:
While I understand your point that we could provide information about the models of the nested objects, I think it would create some inconsistency and unmet expectations, because everywhere else on the application we're only operating on top-level entries (aka "documents"). Even if we start telling users in the CLI that they actually have 20 "models" instead of the 10 we've been presenting them, they won't really be used for anything because we're not treating these nested objects as entries – e.g. they won't generate a call to the `writeFile` method of target plugins.

In my opinion, we should keep the existing behaviour until we expand the API that allows plugins to access the model objects, including fields and their schemas.

Reply:
I agree.

FYI, In Contentful source plugin we never used nested models because there could not be any. In Sanity the story is a bit different. In any case, as you noted, there is no need now to create deep hierarchy of models. But whenever demand will arise, we will need to make sure we expose all the possible models and their fields, including nested models. In format defined by [uniform](https://docs.stackbit.com/uniform/content-models/)